HealthGPT/HealthGPT/MessageManager.swift

```diff
@@ -0,0 +1,84 @@
+//
+// This source file is part of the Stanford HealthGPT project
+//
+// SPDX-FileCopyrightText: 2023 Stanford University & Project Contributors
+//
+
+import Foundation
+import OpenAI
+
+
+@MainActor
+class MessageManager: ObservableObject {
+    @Published private(set) var messages: [Message]
+    @Published private(set) var isQuerying = false
+    private let openAIManager: OpenAIManager
+    private let healthDataFetcher = HealthDataFetcher()
+
+    /// Initializes a new instance of `MessageManager` with the specified API token and OpenAI model.
+    ///
+    /// - Parameters:
+    ///   - apiToken: The API token for the OpenAI API.
+    ///   - openAIModel: The OpenAI model to use for querying.
+    init(apiToken: String = "", openAIModel: Model = .gpt3_5Turbo) {
+        self.messages = []
+
+        self.openAIManager = OpenAIManager(
+            apiToken: apiToken,
+            openAIModel: openAIModel
+        )
+    }
+
+    /// Updates the API token for the OpenAI API.
+    ///
+    /// - Parameter newToken: The new API token.
+    func updateAPIToken(_ newToken: String) {
+        self.openAIManager.updateAPIToken(newToken)
+    }
+
+    /// Updates the OpenAI model to use for querying.
+    ///
+    /// - Parameter newModel: The new OpenAI model to use.
+    func updateOpenAIModel(_ newModel: Model) {
+        self.openAIManager.updateModel(newModel)
+    }
+
+    /// Processes the user message by appending it to the messages list and sending a query to the OpenAI API.
+    ///
+    /// - Parameter userMessage: The user message to process.
+    func processUserMessage(_ userMessage: String) async {
+        let newMessage = Message(content: userMessage, isBot: false)
+
+        self.messages.append(newMessage)
+
+        do {
+            let healthData = try await healthDataFetcher.fetchAndProcessHealthData()
+
+            let generator = PromptGenerator(with: healthData)
+            let mainPrompt = generator.buildMainPrompt()
+
+            Task {
```

Comment:
A more meta-level comment here: I would either put all the work in a single task to perform everything within the task context, making this function itself not async (basically just kicking off a task), or remove the task and also await the response from OpenAI.

 If we do the second option, we could, e.g., enable UI elements to show the progress more easily as they can change their state before and after the async call to this function.

Reply:
Thank you for the feedback! I have removed the Task as you suggested: [e1c0658](https://github.com/StanfordBDHG/HealthGPT/pull/23/commits/e1c06584c1c26505feb21e61497a0293f08de804)